src/controller/CHIPDeviceController.h

```diff
@@ -148,6 +150,8 @@ class DLL_EXPORT DeviceController : public SecureSessionMgrDelegate, public Pers
 
     CHIP_ERROR SetUdpListenPort(uint16_t listenPort);
 
+    chip::app::CommandSender * GetCommandSender() { return mCommandSender; }
```

Comment:
The non-deprecated `ChipDeviceController` API does not send messages directly. Instead it relies on `ChipDevice`, which itself exposes a `SendMessage` method.

I would like to see `GetCommandSender` integrated with `ChipDevice` in some ways instead of bypassing it.

I understand that the current `ChipDevice` API may conflicts with the requirement of the IM, and imho this is the first thing that we need to fix in order to integrate your changes properly. 

It will makes it clearer what is the plan to integrate those changes taking into account the existing codebase.

Reply:
The command sender is not bind to any device, to send a im command, it will pass the command to interaction model->exchange manager->secure session manager->transport.
In this case, does it makes sense for you if add `SendCommand(CommandSender *)` to the ChipDevice class?

Reply:
I would prefer if `ChipDevice` abstract the `CommandSender` inside `SendCommand`. 
`ChipDevice` can then call `CommandSender::SendCommandRequest(mDeviceId)` as it is the one with the relevant `nodeId`

As you may noticed `ChipDevice` also contains a `OnMessageReceived` method, which I guess is where you want some interaction with the `CommandSender::OnMessageReceived`  part

Reply:
ChipDevice is mainly used for OOBE/pairing. IM can be used after OOBE/pairing, and can send message to different kind of devices via local connection, and even mobile vs service in the future, so the command sender is not bind to any device.